### ISSUE TITLE
Unicode support for keys

### ISSUE THREAD

**Issue by abstrctn** [newsdev/archieml.org#3]
[@dr_tj](https://twitter.com/dr_tj) asked whether Unicode support was coming for keys. At the moment, we allow only [-_A-Za-z0-9].

The two most apparent examples would be latin-character names with accented characters, and non-latin words.

In order to support this, we first have to decide what range of characters we'd want to support. If that's everything, what complications might arise in the parsers to support that?

**Comment by abstrctn** [comment]
Relevant: question on whether to allow `@` and possibly `:` in keys: newsdev/archieml-js#2

**Comment by bollwyvl** [comment]
Bringing my thoughts over from newsdev/archieml-js#2.

An AML document could be interpreted as a piece of rich linked data by treating the generated JSON as [JSON-LD](http://json-ld.org). While JSON-LD is developer-friendly, it's not really user-friendly... so many quotes and braces and nesting! AML would fix this.

While this can be done entirely [out-of-band](http://www.w3.org/TR/json-ld/#interpreting-json-as-json-ld), being able to specify an explicit context would be really valuable.

``` yaml
name: George
```

becomes:

``` yaml
@context: http://schema.org/docs/jsonldcontext.json.txt
name: George
```

And now `name` can be unambiguously interpreted as an instance of `http://schema.org/name`, about the node described in this document.

The spec also supports namespaced values (like XML) and full absolute IRIs, both of which would require accepting colons, slashes, dots, etc.

This is not parser-friendly, for obvious reasons.

This is not user friendly, as one would always be better off crafting a more complete context... which can be handled entirely in the context document. However, since left-hand-side values can contain these characters, bringing in a new namespace is still pretty easy:

``` yaml
@id: Alice
[knows]
* bob

{@context}
foaf: http://xmlns.com/foaf/0.1/
{@context.knows}
@id: foaf:knows
@type: @id
```

Finally, as an added bonus, JSON-LD provides an internationalization scheme (which, again, could be hidden in an external context):

``` yaml
@id: Carol
{jobTitle}
en: Doctor
de: Ärtzt

{@context}
sdo: http://schema.org/
{@context.jobTitle}
@id: sdo:jobTitle
@container: @lang
```

**Comment by abstrctn** [comment]
Looking at other languages that allow broad unicode use in variable assignment (such as Swift), I came up with a blacklist of certain characters we should probably avoid.

This doesn't go as far as some other languages: for example, I'm not including mathematical symbols, or punctuation in general. The "ASCII Punctuation & Symbols" table in unicode, which is a superset of most of these reserved characters, contains others that could potentially be useful. `_` is included in that set, as are symbols like `$` which could be useful in representing things like jQuery.

whitespace
- 0000 null
- 0009 horizontal tab
- 000A line feed
- 000B vertical tab
- 000C form feed
- 000D carriage return
- 0020 space
- FEFF BOM

AML reserved characters
- 005C \
- 005B [
- 005D ]
- 007B {
- 007D }
- 003A :

Allowed in slug matching, but interpreted outside of the slug
- 002E .
- 002B + (in use in the freeform markup branch; possibly only when leading, or trailing?)

Possibly reserve for future use?
- 005E ^
- 0060 `
- 007C |
- 007E ~

I have a [feature branch](https://github.com/newsdev/archieml-js/tree/feature/unicode) on `archieml-js` with this implemented.

**Comment by abstrctn** [comment]
In response to a question about using unicode key support in combination with freeform arrays, with regards to being able to have keys such as `span.class` or `span#id`:

The freeform/unicode branch should support '#' in any key, but it looks like it will also support '.' in keys within a freeform array.

While freeforms support opening nested objects with a '{.namespace}' line, the shortcut used in vanilla aml ('namespace.key') doesn't work within freeform blocks, and the '.' just gets preserved as part of the key. Enter this into the sandbox link below to see the different behaviors:

http://archieml.org/sandbox.html?branch=feature/unicode

```
[+ff]
key: value
.key: value
sub.key: value
{.sub}
key: value
{}
#key: value
span#key: value
[]
```

While this wasn't intentional in the implementation, I think it may make sense from a usage perspective as well: every complete statement within a freeform array becomes a single item in that array, and later lines in a freeform don't modify earlier lines (opening a nested object means everything in that object belongs in a single array item, but that's a deliberate grouping).

But using the 'namespace.key' shortcut multiple times within a freeform would mean building up a merged object over multiple lines, which is conceptually quite different, and so maybe the "bug" of periods passing into the keys is desired.

tl;dr, I think we can use class and id CSS selectors out of the box within freeform arrays with the unicode branch?
